### ISSUE TITLE
Bug when torch.topk called on nan's, produces strange results

### ISSUE THREAD

**Issue by epwalsh** [allenai/allennlp#2028]
**Describe the bug**
This is due to this known inconsistency on PyTorch: https://github.com/pytorch/pytorch/issues/1810. I noticed this bug when using beam search on a validation set, after the loss had exploded. Due to the loss exploding, the `step` function being passed to `BeamSearch` was producing `nan`'s for the log probabilities, and then when the beam search called the `topk` function, the top indices ending up being massive values on the order of `9e+18`.. way out of bounds obviously which caused issues downstream.

This isn't really a bug with anything in AllenNLP, since the model shouldn't be passing in `nan` values to BeamSearch, but I still think BeamSearch should raise an error in this case so the issue is caught immediately. (This took me a couple hours to debug because I was unaware of this issue with PyTorch, and therefore didn't realize it was stemming from the output of beam search)

**To Reproduce**
Use `torch.topk` on a tensor filled with `nan` values on GPU.

**Expected behavior**
I think BeamSearch should raise error instead of returning indices that are out of bound. Maybe it's worth having a helper function in `nn.utils` for a "safe topk" function, because I'm sure `topk` is being called in a lot of other places besides `nn.beam_search`. I'd be willing to submit a PR here depending on what you think the best way to handle this is.

**Comment by epwalsh** [comment]
Related to this issue, I've always thought a training loop should raise an error when it encounters nan loss. Is there a good reason not to?

**Comment by epwalsh** [comment]
I'll do a quick benchmark for that. But either way I suppose catching `nan` in the loss would be good to have. I can add a PR for that at least if you have no objections.
